enumerables.rb

```diff
@@ -0,0 +1,203 @@
+# rubocop:disable Metrics/PerceivedComplexity,Metrics/CyclomaticComplexity,Metrics/ModuleLength
+#  rubocop:disable Metrics/MethodLength,Style/CaseEquality
+module Enumerable
+  def my_each
+    return to_enum(:my_each) unless block_given?
+
+    each do |item|
+      yield item
+    end
+
+    self
+  end
+
+  def my_each_with_index
+    return to_enum(:my_each_with_index) unless block_given?
+
+    index = 0
+    each do |item|
+      yield item, index
+      index += 1
+    end
+
+    self
+  end
+
+  def my_select
+    return to_enum(:my_select) unless block_given?
+
+    results_array = []
+    each do |item|
+      next unless yield item
+
+      results_array << item
+    end
+    results_array
+  end
+
+  def my_all?(arg = nil)
+    result = true
+    if block_given?
+      each do |item|
+        result = false unless yield item
+      end
+    elsif arg
+      if arg.is_a?(Regexp)
+        each do |item|
+          result = false unless arg.match(item)
+        end
+      else
+        each do |item|
+          result = false unless arg === item
+        end
+      end
+    else
+      each do |item|
+        result = false unless item
+      end
+    end
+    result
+  end
+
+  def my_any?(arg = nil)
+    result = false
+    if block_given?
+      each do |i|
+        result = true if yield i
+
+        next
+      end
+    elsif arg
+      if arg.is_a?(Regexp)
+        each do |item|
+          result = true if arg.match(item)
+        end
+      else
+        each do |item|
+          result = true if arg === item
+        end
+      end
+    else
+      each do |i|
+        result = true if i
+
+        next
+      end
+    end
+    result
+  end
+
+  def my_none?(arg = nil)
+    result = true
+    if block_given?
+      each do |item|
+        result = false if yield item
+      end
+    elsif arg
+      if arg.is_a?(Regexp)
+        each do |item|
+          result = false if arg.match(item)
+        end
+      else
+        each do |item|
+          result = false if arg == item
+        end
+      end
+    else
+      each do |item|
+        result = false if item
+      end
+    end
+    result
+  end
+
+  def my_count(number = nil)
+    input = Array(self)
+
+    count = 0
+    if number
+      each do |i|
+        count += 1 if i == number
+      end
+    else
+      return input.length unless block_given?
+
+      each do |i|
+        count += 1 if yield i
+      end
+    end
+    count
+  end
+
+  def my_map(proc = nil)
+    end_array = []
+    item = nil
+    each do |i|
+      if proc
+        item = proc.call(i)
+      else
+        return to_enum(:my_map) unless block_given?
+
+        item = yield i
+      end
+      end_array.push item
+    end
+    end_array
+  end
+
+  def my_inject(*args)
+    acc = 0
+    operation = nil
+    if args.empty?
+      return raise LocalJumpError, 'no block given' unless block_given?
+
+      acc = Array(self)[0]
+      each do |i|
+        acc = yield(i, acc)
```

Comment:
# my inject:

- [ ] when a block is given without an initial value, combines all elements of enum by applying a binary operation, specified by a block:

```ruby
let(:operation) { proc { |sum, n| sum + n } }
[1,2,3,4,3,4,2,6,7,4].my_inject(&operation) === [1,2,3,4,3,4,2,6,7,4].inject(&operation) => false `
```
```ruby
Range.new(5, 50).my_inject { |prod, n| prod * n } === Range.new(5, 50).inject { |prod, n| prod * n } => false
```

- [ ] when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method:


```ruby
Range.new(5, 50).my_inject {:*} === Range.new(5, 50).inject {:*} => false
```